driver-embedded-android/src/androidTest/java/com/mongodb/embedded/client/JsonPoweredCrudTestHelper.java

```diff
@@ -0,0 +1,816 @@
+        }
+        return options;
+    }
+
+    private ReplaceOptions getReplaceOptions(final BsonDocument requestArguments) {
+        ReplaceOptions options = new ReplaceOptions();
+        if (requestArguments.containsKey("upsert")) {
+            options.upsert(true);
+        }
+        if (requestArguments.containsKey("collation")) {
+            options.collation(getCollation(requestArguments.getDocument("collation")));
+        }
+        return options;
+    }
+
+    @Nullable
+    private List<BsonDocument> getArrayFilters(@Nullable final BsonArray bsonArray) {
+        if (bsonArray == null) {
+            return null;
+        }
+        List<BsonDocument> arrayFilters = new ArrayList<BsonDocument>(bsonArray.size());
+        for (BsonValue cur : bsonArray) {
+            arrayFilters.add(cur.asDocument());
+        }
+        return arrayFilters;
+    }
+
+    private MongoCollection<BsonDocument> getCollection(final BsonDocument collectionOptions) {
+        MongoCollection<BsonDocument> retVal = baseCollection;
+        if (collectionOptions.containsKey("readPreference")) {
+            retVal = retVal.withReadPreference(getReadPreference(collectionOptions));
+        }
+
+        if (collectionOptions.containsKey("writeConcern")) {
+            WriteConcern writeConcern = getWriteConcern(collectionOptions);
+            retVal = retVal.withWriteConcern(writeConcern);
+        }
+
+        if (collectionOptions.containsKey("readConcern")) {
+            ReadConcern readConcern = getReadConcern(collectionOptions);
+            retVal = retVal.withReadConcern(readConcern);
+        }
+
+        return retVal;
+    }
+
+    ReadPreference getReadPreference(final BsonDocument arguments) {
+        return ReadPreference.valueOf(
+                arguments.getDocument("readPreference").getString("mode").getValue());
+    }
+
+    WriteConcern getWriteConcern(final BsonDocument arguments) {
+        BsonDocument writeConcernDocument = arguments.getDocument("writeConcern");
+        if (writeConcernDocument.size() > 1) {
+            throw new UnsupportedOperationException("Write concern document contains unexpected keys: " + writeConcernDocument.keySet());
+        }
+        if (writeConcernDocument.isNumber("w")) {
+            return new WriteConcern(writeConcernDocument.getNumber("w").intValue());
+        } else {
+            return new WriteConcern(writeConcernDocument.getString("w").getValue());
+        }
+    }
+
+    ReadConcern getReadConcern(final BsonDocument arguments) {
+        return new ReadConcern(ReadConcernLevel.fromString(arguments.getDocument("readConcern").getString("level").getValue()));
+    }
+
+
+    public static BsonDocument getTestDocument(final String json) {
+        return new BsonDocumentCodec().decode(new JsonReader(json), DecoderContext.builder().build());
+    }
+
+    public static List<BsonDocument> getTestDocuments(final String resourcePath) {
```

Comment:
This is the main difference to `util/JsonPoweredTestHelper` - I ended up copying all the code into Android Studio as a standalone project to be able to debug the api correctly and get this to work.

Reply:
What a PITA